File: ma_sh/Cpp/src/sample.cpp
#include "sample.h"
#include "constant.h"
#include <c10/core/DeviceType.h>
#include <cstdint>
#include <torch/types.h>

using namespace torch::indexing;

const torch::Tensor toUniformSamplePhis(const int &sample_num) {
  std::vector<float> phis_vec;
  phis_vec.reserve(sample_num);

  for (int i = 0; i < sample_num; ++i) {
    phis_vec.emplace_back(PHI_WEIGHT * (i + 0.5));
  }

  const torch::TensorOptions opts =
      torch::TensorOptions().dtype(torch::kFloat32).device(torch::kCPU);

  const torch::Tensor phis =
      torch::from_blob(phis_vec.data(), {sample_num}, opts).clone();

  return phis;
}

const torch::Tensor toUniformSampleThetas(const int &sample_num) {
  std::vector<float> cos_thetas_vec;
  cos_thetas_vec.reserve(sample_num);

  for (int i = 0; i < sample_num; ++i) {
    cos_thetas_vec[i] = 1.0 - (2.0 * i + 1.0) / sample_num;
  }

  const torch::TensorOptions opts =
      torch::TensorOptions().dtype(torch::kFloat32).device(torch::kCPU);

  const torch::Tensor cos_thetas =
      torch::from_blob(cos_thetas_vec.data(), {sample_num}, opts).clone();

  const torch::Tensor thetas = torch::acos(cos_thetas);

  return thetas;
}

const torch::Tensor toMaskBoundaryPhis(const int &anchor_num,
                                       const int &mask_boundary_sample_num) {
  torch::Tensor mask_boundary_phi_matrix =
      torch::zeros({anchor_num, mask_boundary_sample_num});

  for (int i = 0; i < mask_boundary_sample_num; ++i) {
    const float current_phi = PI_2 * i / mask_boundary_sample_num;

    mask_boundary_phi_matrix.index_put_({slice_all, i}, current_phi);
  }

  const torch::Tensor mask_boundary_phis =
      mask_boundary_phi_matrix.reshape({-1});

  return mask_boundary_phis;
}

const torch::Tensor toSampleThetaNums(const torch::Tensor &mask_boundary_thetas,
                                      const float &delta_theta_angle) {
  const torch::Tensor detach_mask_boundary_thetas =
      mask_boundary_thetas.detach();

  const float delta_theta = M_PI / 90.0 * delta_theta_angle;

  const torch::Tensor sample_theta_nums =
      torch::ceil(detach_mask_boundary_thetas / delta_theta)
          .toType(torch::kInt64);

  return sample_theta_nums;
}

const torch::Tensor toSampleThetas(const torch::Tensor &mask_boundary_thetas,
                                   const torch::Tensor &sample_theta_nums) {
  std::vector<torch::Tensor> sample_thetas_vec;
  sample_thetas_vec.reserve(sample_theta_nums.size(0));

  const torch::TensorOptions opts = torch::TensorOptions()
                                        .dtype(mask_boundary_thetas.dtype())
                                        .device(mask_boundary_thetas.device());

  for (int i = 0; i < sample_theta_nums.size(0); ++i) {
    const int current_sample_theta_num = sample_theta_nums[i].item<int>();

    const torch::Tensor current_sample_thetas =
        torch::arange(1, current_sample_theta_num + 1, opts) /
        current_sample_theta_num * mask_boundary_thetas[i];

    sample_thetas_vec.emplace_back(current_sample_thetas);
  }

  const torch::Tensor sample_thetas = torch::hstack(sample_thetas_vec);

  return sample_thetas;
}

File: ma_sh/Method/Mash/sample.py
import torch

from ma_sh.Config.constant import PI_2, PHI_WEIGHT


def toUniformSamplePhis(sample_num: int) -> torch.Tensor:
    phis_list = [PHI_WEIGHT * (i + 0.5) for i in range(sample_num)]

    phis = torch.tensor(phis_list)

    return phis


def toUniformSampleThetas(sample_num: int) -> torch.Tensor:
    cos_thetas_list = []
    for i in range(sample_num):
        cos_thetas_list.append(1.0 - (2.0 * i + 1.0) / sample_num)

    cos_thetas = torch.tensor(cos_thetas_list)

    thetas = torch.acos(cos_thetas)

    return thetas


def toMaskBoundaryPhis(anchor_num: int, mask_boundary_sample_num: int) -> torch.Tensor:
    mask_boundary_phi_matrix = torch.zeros(anchor_num, mask_boundary_sample_num)

    for i in range(mask_boundary_sample_num):
        mask_boundary_phi_matrix[:, i] = PI_2 * i / mask_boundary_sample_num

    mask_boundary_phis = mask_boundary_phi_matrix.reshape(-1)

    return mask_boundary_phis


def toFPSIdxs(points: torch.Tensor, fps_point_num: int) -> torch.Tensor:
    """
    Input:
        points: pointcloud data, [B, N, C]
        fps_point_num: number of samples
    Return:
        fps_point_idxs: sampled pointcloud index, [B, fps_point_num]
    """
    device = points.device
    B, N, C = points.shape

    centroids = torch.zeros(B, fps_point_num, dtype=torch.long).to(device)

    distance = torch.ones(B, N).to(device) * 1e10

    farthest = torch.randint(0, N, (B,), dtype=torch.long).to(device)

    batch_indices = torch.arange(B, dtype=torch.long).to(device)
    for i in range(fps_point_num):
        centroids[:, i] = farthest
        centroid = points[batch_indices, farthest, :].view(B, 1, 3)
        dist = torch.sum((points - centroid) ** 2, -1)
        mask = dist < distance
        distance[mask] = dist[mask]
        farthest = torch.max(distance, -1)[1]
    return torch.sort(centroids)[0]
